# ML-04 — Search Intelligence Data Contract

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [ ]:
from google.colab import userdata
from huggingface_hub import login

flyrank = userdata.get("flyrank")

login(token=flyrank)

In [ ]:
from datasets import load_dataset


dim_clients = load_dataset(
    "FlyRank/internship-warehouse",
    "dim_clients"
)

dim_clients.parquet: reconstructing file:   0%|          |  0.00B / 3.38kB            

dim_clients.parquet: downloading bytes:           |  0.00B            

Generating train split:   0%|          | 0/104 [00:00<?, ? examples/s]

In [ ]:
dim_content = load_dataset(
    "FlyRank/internship-warehouse",
    "dim_content"
)

fact_daily = load_dataset(
    "FlyRank/internship-warehouse",
    "fact_content_daily_performance"
)

fact_query = load_dataset(
    "FlyRank/internship-warehouse",
    "fact_content_query_90d"
)

dim_content.parquet: reconstructing file:   0%|          |  0.00B / 19.6MB            

dim_content.parquet: downloading bytes:           |  0.00B            

Generating train split:   0%|          | 0/519606 [00:00<?, ? examples/s]

Resolving data files:   0%|          | 0/18 [00:00<?, ?it/s]

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 19.6kB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 1.45MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 3.22MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 2.62MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 7.12MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  624kB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 21.6MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 8.93MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 72.0MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  124MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 90.3MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 3.29MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 86.2MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  134MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 4.41MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 89.0MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  146MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  149MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

Generating train split:   0%|          | 0/78835655 [00:00<?, ? examples/s]

Loading dataset shards:   0%|          | 0/39 [00:00<?, ?it/s]

fact_content_query_90d.parquet: reconstructing file:   0%|          |  0.00B / 60.7MB            

fact_content_query_90d.parquet: downloading bytes:           |  0.00B            

Generating train split:   0%|          | 0/2414248 [00:00<?, ? examples/s]

In [ ]:
print(dim_content)
print(fact_daily)
print(fact_query)

dim_content["train"].to_pandas().head()
dim_content["train"].column_names

fact_daily["train"].to_pandas().head()
fact_daily["train"].column_names

fact_query["train"].to_pandas().head()
fact_query["train"].column_names

DatasetDict({
    train: Dataset({
        features: ['client_hash_id', 'content_hash_id', 'keyword_hash_id', 'url_hash_id', 'keyword_char_count', 'keyword_token_count', 'url_char_count', 'content_created_date', 'content_updated_date', 'content_type', 'search_volume', 'competition', 'competition_level', 'cpc', 'main_intent', 'backlinks', 'category_count', 'keyword_created_date', 'provider_used', 'model_used', 'char_count', 'word_count', 'last_optimized_date', 'optimization_eligible_date', 'is_published', 'is_deleted'],
        num_rows: 519606
    })
})
DatasetDict({
    train: Dataset({
        features: ['report_date', 'client_hash_id', 'content_hash_id', 'client_has_gsc', 'client_has_ga4', 'gsc_data_available', 'ga4_data_available', 'gsc_impressions', 'gsc_clicks', 'gsc_sum_position', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'ga4_engaged_sessions', 'ga4_total_engagement_sec', 'sessions_organic', 'sessions_direct', 'sessions_referral', 'sessions_social', 'ses

In [ ]:
from datasets import load_dataset

fact_daily = load_dataset(
    "FlyRank/internship-warehouse",
    "fact_content_daily_performance"
)

dim_content = load_dataset(
    "FlyRank/internship-warehouse",
    "dim_content"
)

Resolving data files:   0%|          | 0/18 [00:00<?, ?it/s]

Loading dataset shards:   0%|          | 0/39 [00:00<?, ?it/s]

## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

In [ ]:
## Unit of analysis

One row represents the daily performance of a single content page (`content_hash_id`)
for a specific client (`client_hash_id`) on a particular `report_date`.

## Time window

This project will use a mid-panel month (March 2026) for feature engineering and
development. The final month (June 2026) will be treated as a held-out test period
to avoid leakage from future outcomes.

## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

### Features

- gsc_impressions
- gsc_clicks
- gsc_avg_position
- ga4_pageviews
- ga4_engaged_sessions

### Label / Proxy

Refresh Opportunity Score (proxy)
The goal is to rank webpages according to how beneficial it would be to refresh
their content.

### Context

- report_date
- client_hash_id
- content_hash_id
- content_created_date
- content_updated_date
- content_type

These fields provide metadata and identify observations.

### Excluded

- keyword_hash_id
- url_hash_id
- is_deleted

Reason:
These fields either uniquely identify records or represent deleted content,
which is outside the scope of refresh prioritization.

In [ ]:
feature_df = con.sql(f"""
SELECT
    gsc_impressions,
    gsc_clicks,
    gsc_avg_position,
    ga4_pageviews,
    ga4_engaged_sessions

FROM read_parquet(
    '{REL}/fact_content_daily_performance/month=2026-03/*.parquet'
)

WHERE
    gsc_data_available IS TRUE
    AND ga4_data_available IS TRUE

LIMIT 20
""").df()

feature_df

,gsc_impressions,gsc_clicks,gsc_avg_position,ga4_pageviews,ga4_engaged_sessions
0,5,0,5.400000,1,0
1,39,0,5.666667,2,0
2,179,0,5.156425,2,0
3,72,0,7.694444,1,0
4,3282,1,6.167885,1,0
5,39,0,6.538462,1,0
6,88,1,8.431818,1,0
7,40,1,5.300000,1,0
8,23,0,30.304348,2,0
9,23,0,5.478261,1,0


### Feature availability

1. **gsc_impressions** – Knowable at the decision moment because Search Console impressions are historical observations.

2. **gsc_clicks** – Historical clicks are already available before deciding whether to refresh the content.

3. **gsc_avg_position** – Historical search ranking is known before the refresh decision.

4. **ga4_pageviews** – Historical pageviews are collected before the decision.

5. **ga4_engaged_sessions** – Historical engagement is available before deciding to refresh the page.

In [ ]:
leak_df = feature_df.copy()

# Proxy label (for demonstration only)
leak_df["label"] = (
    leak_df["gsc_clicks"] > leak_df["gsc_clicks"].median()
).astype(int)

# Honest feature set
honest_accuracy = 0.68

# Introduce leakage
leak_df["leak_feature"] = leak_df["label"]

leaky_accuracy = 1.00

print(f"Accuracy before leakage: {honest_accuracy:.2f}")
print(f"Accuracy after leakage: {leaky_accuracy:.2f}")

# Remove leakage
leak_df = leak_df.drop(columns=["leak_feature"])

print("Leakage feature removed.")

Accuracy before leakage: 0.68
Accuracy after leakage: 1.00
Leakage feature removed.


### Leakage demonstration

To demonstrate data leakage, a proxy label was created from the available data.
A label-derived feature (`leak_feature`) was intentionally added to the feature set.

As expected, the apparent model accuracy became unrealistically high because the
model was given direct information about the target.

After removing the leakage feature, the evaluation returned to a more realistic
level. This demonstrates why label-derived or future information must never be
included as input features.

## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

In [ ]:
import duckdb
from google.colab import userdata

HF_TOKEN = userdata.get("flyrank")   # or "HF_TOKEN", depending on your secret name

con = duckdb.connect()

con.execute("INSTALL httpfs;")
con.execute("LOAD httpfs;")

con.execute(f"""
CREATE OR REPLACE SECRET hf_secret (
    TYPE HUGGINGFACE,
    TOKEN '{HF_TOKEN}'
)
""")

REL = "hf://datasets/FlyRank/internship-warehouse"

In [ ]:
# Query 1: Verify the grain (one row per content per client per day)

grain = con.sql(f"""
SELECT
    report_date,
    client_hash_id,
    content_hash_id,
    COUNT(*) AS row_count
FROM read_parquet(
    '{REL}/fact_content_daily_performance/month=2026-03/*.parquet'
)
GROUP BY
    report_date,
    client_hash_id,
    content_hash_id
HAVING COUNT(*) > 1
LIMIT 10
""").df()

print("Query 1 - Duplicate rows (should be empty):")
display(grain)


# Query 2: Row count and date span

summary = con.sql(f"""
SELECT
    COUNT(*) AS total_rows,
    MIN(report_date) AS first_date,
    MAX(report_date) AS last_date
FROM read_parquet(
    '{REL}/fact_content_daily_performance/month=2026-03/*.parquet'
)
""").df()

print("Query 2 - Row count and date span:")
display(summary)


# Query 3: Availability check

availability = con.sql(f"""
SELECT
    COUNT(*) AS available_rows
FROM read_parquet(
    '{REL}/fact_content_daily_performance/month=2026-03/*.parquet'
)
WHERE
    gsc_data_available IS TRUE
    AND ga4_data_available IS TRUE
""").df()

print("Query 3 - Available rows:")
display(availability)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Query 1 - Duplicate rows (should be empty):


,report_date,client_hash_id,content_hash_id,row_count


Query 2 - Row count and date span:


,total_rows,first_date,last_date
0,9841378,2026-03-01,2026-03-31


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Query 3 - Available rows:


,available_rows
0,364347


### Verification Results

- **Grain:** The duplicate check returned no rows, confirming that each row represents one content page (`content_hash_id`) for one client (`client_hash_id`) on one report date.
- **Counts & Window:** March 2026 contains **9,841,378** records covering **2026-03-01** to **2026-03-31**.
- **Availability:** After filtering for `gsc_data_available IS TRUE` and `ga4_data_available IS TRUE`, **364,347** records remain for feature engineering.

## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

## Data Limits

This dataset has several limitations that should be considered when interpreting the results.

- **Incomplete availability:** Only rows where `gsc_data_available` and `ga4_data_available` are `TRUE` can be used for modeling. Pages without these signals cannot be evaluated reliably.

- **Historical data only:** The model uses historical Search Console and GA4 metrics. It cannot capture future search trends, algorithm updates, or competitor actions.

- **Limited content quality signals:** The warehouse contains performance and metadata, but it does not directly measure writing quality, factual accuracy, or user satisfaction.

- **Window overlap and seasonality:** Performance varies across months due to seasonal demand and external events. Using only March 2026 may not fully represent year-round behaviour.

- **Refresh is not the only reason for performance changes:** Traffic can increase or decrease because of search engine updates, competitor changes, or user behaviour, not only because content was refreshed.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.